init/deviceref.go

```diff
@@ -86,24 +88,87 @@ func (d *deviceRef) resolveFromGptTable(devPath string, t []gptPart) *deviceRef
 	return d
 }
 
-var autodiscoveryGptTypes = map[string]string{
+var autodiscoveryRootGptTypes = map[string]string{
 	"amd64": "4f68bce3-e8cd-4db1-96e7-fbcaf984b709",
 	"386":   "44479540-f297-41b2-9af7-d131d5f0458a",
 	"arm":   "69dad710-2ce4-4e3c-b16c-21a1d49abed3",
 	"arm64": "b921b045-1df0-41c3-af44-4c6f280d3fae",
 	//"itanium": "993d8d3d-f80e-4225-855a-9daf8ed7ea97",
 }
 
+func (d *deviceRef) autodetectRoot(devPath string, t []gptPart) *deviceRef {
+	rootTypeGuidRaw, ok := autodiscoveryRootGptTypes[runtime.GOARCH]
+	if !ok {
+		return d
+	}
+	rootTypeGuid, err := parseUUID(rootTypeGuidRaw)
+	if err != nil {
+		return d
+	}
+
+	debug("Scanning %s for the root partition. Used ESP was found in this device.", devPath)
+
+	for _, p := range t {
+		if bytes.Equal(rootTypeGuid, p.typeGuid) {
+			// TODO: Honour GPT flags (read only & no auto)
+			rootDevName := calculateDevName(devPath, p.num)
+			debug("Found root partition: %s.", rootDevName)
+			info, err := readBlkInfo(rootDevName)
+			if err == nil {
```

Comment:
this condition should be flipped to make the error handling more go-idiomatic:

if err != nil {
  debug("foo")
  continue
}